Add Docathon documentation (#172)
stackabletech/documentation: modules/contributor/pages/guides/docathon-22-01/druid.adoc

= Overview

> This section describes how to set up a streaming job in Druid to ingest data from the Kafka topic that was used with Nifi. Druid requires a backing storage (so called Deep-Storage) where data - partioned by date or time - is persisted as immutable "segments". Druid can use either local storage (only appropriate for stand-alone testing - i.e. all druid components run on the same machine), S3 or HDFS. In this guide we will use S3, specifically Min-IO which is an S3-implementation suitable for low-footprint scenarios.

Reviewer: ```suggestion
This section describes how to set up a streaming job in Druid to ingest data from the Kafka topic that was used with Nifi. Druid requires a backing storage (so called Deep-Storage) where data - partioned by date or time - is persisted as immutable "segments". Druid can use either local storage (only appropriate for stand-alone testing - i.e. all druid components run on the same machine), S3 or HDFS. In this guide we will use S3, specifically MinIO which is an S3-implementation suitable for low-footprint scenarios.
```

> == Deploy Min-IO

Reviewer: ```suggestion
== Deploy MinIO
```

> [source]

Reviewer: ```suggestion
[source,bash]
```

helm install minio --set resources.requests.memory=8Gi --set mode=standalone --set replicas=1  --set persistence.enabled=false  --set "buckets[0].name=nytaxidata,buckets[0].policy=none" --set "users[0].accessKey=minioAccessKey,users[0].secretKey=minioSecretKey,users[0].policy=readwrite" --repo https://charts.min.io/ minio

N.B.

- we are specifying a memory allocation of 8GB as Min-IO will use 16GB by default.
- the access credentials are given above in the form of a secret, which will be used later in the Druid custom resource.

== Deploy Postgresql (for Druid metadata)

Druid requires a database to store metadata: for this guide we will use the Bitnami PostgreSQL helm chart to deploy a PostgreSQL instance:

> [source]

Reviewer: ```suggestion
[source,bash]
```

helm install postgresql-druid \
    --repo https://charts.bitnami.com/bitnami postgresql \
    --set auth.username=druid \
    --set auth.password=druid \
    --set auth.database=druid \
    --version 11.0.0

N.B. the credentials will also be used in the Druid custom resource.

== Deploy the Stackable Druid operator

> [source]

Reviewer: ```suggestion
[source,bash]
```

    roleGroups:
      default:
        selector:
          matchLabels:
            kubernetes.io/os: linux
        config: {}
        replicas: 1
  coordinators:
    configOverrides:
      runtime.properties:
        druid.s3.enablePathStyleAccess: "true"
    roleGroups:
      default:
        selector:
          matchLabels:
            kubernetes.io/os: linux
        config: {}
        replicas: 1
  historicals:
    configOverrides:
      runtime.properties:
        druid.s3.enablePathStyleAccess: "true"
    roleGroups:
      default:
        selector:
          matchLabels:
            kubernetes.io/os: linux
        config: {}
        replicas: 1
  middleManagers:
    configOverrides:
      runtime.properties:
        druid.s3.enablePathStyleAccess: "true"
    roleGroups:
      default:
        selector:
          matchLabels:
            kubernetes.io/os: linux
        config: {}
        replicas: 1
  routers:
    configOverrides:
      runtime.properties:
        druid.s3.enablePathStyleAccess: "true"
    roleGroups:
      default:
        selector:
          matchLabels:
            kubernetes.io/os: linux
        config: {}
        replicas: 1
EOF

> <1> s3 secret

Reviewer 2: ```suggestion
<1> S3 secret
```

<2> Druid ZNode

== Data Ingestion

There are different ways to get data into Druid, all of which will use a `POST` of a Druid-compatible ingestion specification. We will document here two ways of doing this, either directly in the Druid UI, or - this is e.g. useful if the job is to be repeated - by extracting the ingestion specification into a JSON file and issuing a curl from the command line (some of what follows is also covered in more depth in the official Druid documentation, but is mentioned here for the sake of completion).

=== Using the Druid UI

==== Setup port-forwarding for the Druid UI

Run this from the command line to open up access to the Druid router (keep this command line tab open):

> [source]

Reviewer: ```suggestion
[source,bash]
```

image::docathon-2022-01/druid-datasources.png[Datasources]

We can also display data by issuing queries against our datasource from within the SQL designer under the “Query” tab:

image::docathon-2022-01/druid-query.png[Query screen]

=== Using `curl`

We will now perform the same action using the JSON specification we saved earlier (in this guide: `/tmp/kafka-ingestion-spec.json`).

==== Setup port-forwarding for Druid Co-ordinator

Issue a port-forwarding command so that we can access the Druid co-ordinator from outside the the Kubernetes cluster:

[source]
kubectl port-forward svc/druid-nytaxidata-coordinator 8081

==== Post Job Specification

Issue a POST via curl, referencing the JSON specification:

[source]

> curl -X POST -H 'Content-Type: application/json' -d @/tmp/kafka-ingestion-spec.json http://localhost:8081/druid/indexer/v1/supervisor

Reviewer: I think we have to add the file `kafka-ingestion-spec.json` to the `downloads` folder

Author: I'd quick like to have the users click through those steps and do it manually first, but you're right, that's probably a good idea - I'm not 100% sure if we have anything dynamic in there that might change. Are you OK with leaving it as it is for the time being?
